### ISSUE TITLE
Field masking

### ISSUE THREAD

**Issue by Nevon** [klarna/ui-react-components#3]
In KCO, we make use of input masking to help the user conform to an expected input format. For example, when we require a date of birth in the US, we ask them to input the date as `MM/DD/YYYY`. So as to help them with this, we automatically add the slashes where appropriate. So if you were to type `12`, we would add the slash and advance the caret so that it says `12/`. Similarly if you now press backspace, we would delete not only the slash but also the first digit after it, so that it would now say `1`.

This is a fairly common use-case. Other cases include CC numbers and PNOs. I think that these cases are common enough within Klarna that it might make sense to bring this into the toolkit in some way.

Any thoughts? Should this be part of the ui toolkit? If so, how can it be done? Should we employ some graceful degradation for troublesome browsers?

--

Some things I learned from having implemented masking in KCO:

* Changing the length of the value in an input in response to an input event screws up the state of it in the old Android stock browser as well as in multiple versions of IE11. Ex. you have an input with the value `1` (length 1), you press `2` and from within the event handler, you also add a slash and attempt to move the caret 2 positions forward. The input value is now `12/` (length 3) and the caret should be at the end of the value. In affected browsers, the caret will be between the 2 and the slash and be unable to move any further to the right, even manually. I don't remember the exact reproduction steps, but it's also very easy to get undeletable characters. This is an extremely common issue with off-the-shelves solutions, likely because they don't have as strict browser support as we do.
* Masking is not just masking. A good experience requires knowledge of what you are masking. For example, in the case of a date of birth, we know that if the format is `MM/DD/YYYY` and the user starts by typing a 4, the only valid value for the month is `04`, so we can actually zero pad and add the slash already. Perhaps there's more than one layer to it. One layer that implements naive masking that only knows about a mask and some value, and another value that knows stuff about the data itself.
* A masked, active field can never be empty, and this means that it will behave differently from normal fields. When you focus a normal field, the placeholder remains until you've typed something. In a masked field, the value of the field becomes the mask as soon as you focus it (otherwise you have no way of knowing the expected format until you've already started typing), so the placeholder will become a label immediately.
* No one has solved this perfectly. Either the user experience, the browser support or the flexibility sucks. Pick your poison.

**Comment by joaomilho** [comment]
I agree about the fact UX is never really smooth. I have seen few designers really thinking about this kind of thing, and if they are "interaction designers" they should study this kinda thing (ok, I shouldn't publicly shame designers, but it's factual :/).

The best thing I've seen is going completely the other way: smart fields. So, in the case of a date you can type "22/01/2000" or "2000-01-22" or "22 of Jan" or "22 January" or "10 years ago"... and the field would either figure out what you mean or ask you to disambiguate (like in e.g. 06/06/06 if it can't figure out the locale of the user).

**Comment by joaomilho** [comment]
Like http://gr2m.github.io/smartdate-input/ and http://laruiss.github.io/smart-date/index.amd.html (still not gr8, but interesting)

**Comment by Nevon** [comment]
That's an interesting approach. I feel like it's more applicable in situations where getting it wrong is not that big of a deal, and where it's obvious to the user if you got it wrong. For example, when I create a calendar event in Google Calendar, I can just type "Lunch @ Stu's Stews with Juliana 12:00 on Monday" and it will pre-fill the calendar event with all the relevant information, but I'll still get to review the information before it's committed - and even if it fails to figure out where "Stu's Stews" is, it's not the end of the world.

Contrast that to getting a date of birth in a credit application wrong.

If you do go that way, I think it's really important that the user can see what it was interpreted at. Perhaps by replacing the value with the interpreted value on blur, or something.

--

For what I think is a fairly good example of date masking, see https://github.com/insin/react-maskedinput.

**Comment by joaomilho** [comment]
I see. But if you look at the example I've sent it combines both approaches.

**Comment by yuval-netanel** [comment]
We are facing a similar issue with myKlarna [new amount field](https://stash.int.klarna.net/projects/KSA/repos/settlement-app/pull-requests/1203/overview) which has a masking. 
What we have decided is to put the component in the myKlarna project, once we will see that it is "mature" enough we will add it this project.

I think that these fields are belong here, we will all benefit from that, by reusing component or just learning from each other mask implementation experience.

**Comment by joaomilho** [comment]
Here's our current implementation, done by @omriyariv https://stash.int.klarna.net/projects/KSA/repos/settlement-app/browse/webpack/components/Amount/index.jsx

I guess it's very number-specific (right to left typing), but we could already have a general number field component based on it. One thing I like about it is that it abstracts the `formatter`, so it will work for many different formats [of numbers].

**Comment by xaviervia** [comment]
Ok, so my question about this is: could we enable the clever masking for supported browsers and fallback to plain input in the other ones?

**Comment by xaviervia** [comment]
Regarding the Date field type, the current standard is to use a combination of native Date type with masking in supported browsers. This solution has been adopted in individual projects but not yet in the UI Components. The aim is to have it ported to the components in the upcoming weeks.

There is no generic solution for masking though.
